### PR TITLE
num_converter

### LINKED ISSUE #21: num_converter
Decimal to binary and vice versa

### PR DESCRIPTION
**Pull request by byabdallah**
---
name: solution review for the num_converter function
about: A code that converts decimal to binary and vice versa
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/tests/test_num_converter.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
+import unittest
```
Comment (hudaalamassi) [inline solutions/tests/test_num_converter.py:14]: (Files/Behavior):

A module-level docstring is required at the very beginning of the file to meet the rule: There is a module docstring in the tests file.

File: solutions/tests/test_num_converter.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+    def test_binary_to_decimal_1111(self):
+        """
+        It should return 15 if the input is 1111
+        """
+        actual = binary_to_decimal(1111)
+        expected = 15
+        self.assertEqual(actual, expected)
+
+    def test_binary_to_decimal_zeros(self):
```
Comment (hudaalamassi) [inline solutions/tests/test_num_converter.py:62]: (Unit tests/Behavior): 

There are tests for boundary cases: This rule is partially fulfilled.

- Suggestion: Add a test for decimal_to_binary(0).
- Zero is a boundary case and is part of testing the lower limits of input.
- Currently, the decimal_to_binary function doesn't explicitly handle the case where decimal == 0. If you pass 0 into the function, it will: 
- Skip the while loop, as the condition decimal != 0 will be False immediately.
- The variable binary_num remains an empty string ("").
- When trying to convert an empty string to an integer, Python will raise a ValueError. This is because int("") is not a valid conversion.

File: solutions/num_converter.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+"""
+
+
+def decimal_to_binary(decimal: int):
+    """
+    Converts a decimal number to its binary representation as a string.
+
+    Parameters:
+        decimal_number (int): The decimal number to be converted.
+
```
Comment (hudaalamassi) [inline solutions/num_converter.py:21]: (Function Docstring/Behavior):

- The docstring for decimal_to_binary(decimal: int) states that the function converts a decimal number to its binary representation as a string, but the function actually returns an integer. I think string should be removed here.

- Other assumptions for the function's arguments: The assumption that the decimal number is an integer is mentioned in the parameters/arguments, but the docstring does not describe the expected range of the integer (e.g., if it should be non-negative).

File: solutions/num_converter.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    return int(binary_num)
+
+
+def binary_to_decimal(binary):
+    """
+    Converts a binary number (as a string or integer) to its decimal representation.
+
+    Parameters:
+        binary (str or int): The binary number to convert.
+
```
Comment (hudaalamassi) [inline solutions/num_converter.py:72]: (Function Docstring/Behavior):

- The description and type rules are not fulfilled for the binary_to_decimal function because:

- The docstring says the input can be either a string or an integer (binary (str or int)), but the function only accepts an integer. This creates a mismatch between the docstring and the actual behavior of the function.

File: solutions/num_converter.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+Module contents:
+    - decimal_to_binary: converts a decimal number into its binary representation.
+    - binary_to_decimal: converts a binary number into its decimal representation.
+
+Created on 31 12 24
+@author: Abdallah Alnajjar
+"""
+
+
+def decimal_to_binary(decimal: int):
```
Comment (hudaalamassi) [inline solutions/num_converter.py]: (The Function/Behavior):

- Type annotations in Python are typically written in the function declaration to specify the expected types of parameters and the return value.

- decimal_to_binary(decimal: int): The type annotation for decimal: int is correct. However, if it’s meant to return an int, the return type should also be specified as -> int in the function declaration.

- binary_to_decimal(binary): This function lacks any type annotations for both the input and the return value. The code need to have them for clarity.

File: solutions/num_converter.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+            binary_list.append(0)
+        decimal = decimal / 2
+
+    # Convert digits array to string in correct order
+    for num in reversed(binary_list):
+        binary_num = binary_num + str(num)
+    return int(binary_num)
+
+
+def binary_to_decimal(binary):
```
Comment (hudaalamassi) [inline solutions/num_converter.py]: (The Function/Behavior):

- This comment just for discussion and I do not think the code need a change here.

- The function's name matches the file name rule: may not apply strictly in this case.

- Multiple Functions: The file contains two functions: decimal_to_binary and binary_to_decimal. Given that the file is designed to handle both binary-to-decimal and decimal-to-binary conversions, it is understandable that the file might not match the name of a specific function.

- In this case, it’s reasonable to ignore the rule that the function's name must match the file name, because the file contains multiple related functions, and the file name (num_converter) likely describes the purpose of the entire file, not just a single function.

File: solutions/num_converter.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    11111110
+    >>> decimal_to_binary(13)
+    1101
+    >>> decimal_to_binary(173)
+    10101101
+
+    """
+
+    # validate input type
+    assert isinstance(decimal, int)
```
Comment (hudaalamassi) [inline solutions/num_converter.py]: Assertion messages could be added to clarify the checks and improve debugging.

File: solutions/num_converter.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    while decimal != 0:
+        if decimal % 2 == 1:
+            binary_list.append(1)
+            decimal -= 1
+        elif decimal % 2 == 0:
+            binary_list.append(0)
+        decimal = decimal / 2
+
+    # Convert digits array to string in correct order
+    for num in reversed(binary_list):
```
Comment (hudaalamassi) [inline solutions/num_converter.py:61]: Strategy:

- Comments explain the strategy, not the implementation: This rule is not completely fulfilled.

- This: # Convert digits array to string in correct order ( more implementation-focus)

- could be : # Reverse the binary digits to form the final binary string 

- The focus of this comment should be on why reversing the binary digits, rather than the specific step of converting the array to a string, which is more of an implementation task.

File: solutions/num_converter.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+    # Ensure the input is treated as a string
+    binary = str(binary)
+    decimal = 0
+
+    # ensure that the input contains only zeros and ones
+    assert all(char in "01" for char in binary)
+
+    # Start writing the binary representation by iterating over each bit.
+    for bit in binary:
```
Comment (hudaalamassi) [inline solutions/num_converter.py:98]: Strategy:

- Comments explain the strategy, not the implementation: This rule is not completely fulfilled.

- This: # Start writing the binary representation by iterating over each bit. ( more implementation-focus)

- could be : # Iterate through the binary string to calculate its decimal equivalent

- The revised comment is more aligned with the strategy, focusing on the goal of iterating over the binary number to compute the corresponding decimal value. The original comment is more about the specific mechanism of how you're doing it (which is fine but more implementation-focused).

File: solutions/num_converter.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    >>> decimal_to_binary(173)
+    10101101
+
+    """
+
+    # validate input type
+    assert isinstance(decimal, int)
+
+    # ensure that the decimal is positive
+    assert decimal >= 0
```
Comment (hudaalamassi) [inline solutions/num_converter.py:42]: Implementation:

decimal: this is a single word, which is acceptable in general but does not fully align with the multi-word naming conventions of snake_case.

File: solutions/num_converter.py (hunk 8 above)
Comment (hudaalamassi) [inline solutions/num_converter.py:61]: Implementation:

num: this is a single word, which is acceptable in general but does not fully align with the multi-word naming conventions of snake_case.

File: solutions/num_converter.py [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+    >>> binary_to_decimal(11111)
+    31
+    """
+    # validate input type
+    assert isinstance(binary, int)
+
+    # Ensure the input is treated as a string
+    binary = str(binary)
+    decimal = 0
+
```
Comment (hudaalamassi) [inline solutions/num_converter.py:93]: Implementation:

binary, decimal: this is a single word, which is acceptable in general but does not fully align with the multi-word naming conventions of snake_case.

File: solutions/num_converter.py (hunk 9 above)
Comment (hudaalamassi) [inline solutions/num_converter.py:98]: Implementation:

bit: this is a single word, which is acceptable in general but does not fully align with the multi-word naming conventions of snake_case.

**Review by hudaalamassi — COMMENTED** [review]
I reviewed the solution and commented under each part of the code.

File: solutions/num_converter.py (hunk 7 above)
Comment (byabdallah) [reply solutions/num_converter.py]: I don't understand this comment

File: solutions/num_converter.py (hunk 10 above)
Comment (byabdallah) [reply solutions/num_converter.py:42]: I don't understand this too. it's a one word, how is it supposed to follow the pattern "snake_case"?

File: solutions/num_converter.py (hunk 8 above)
Comment (byabdallah) [reply solutions/num_converter.py:61]: Implementation is when you describe the code. For example "This is a for loop that iterates over the reversed version of binary_list and then stores the values in a variable called binary_num". This comment is not literally describing the function of the code. However, this comment will be changed.

Comment (byabdallah) [reply solutions/num_converter.py:61]: Again, I don't understand this too. it's a one word, how is it supposed to follow the pattern "snake_case"?

File: solutions/num_converter.py (hunk 11 above)
Comment (byabdallah) [reply solutions/num_converter.py:93]: Again, I don't understand this too. it's a one word, how is it supposed to follow the pattern "snake_case"?

File: solutions/num_converter.py (hunk 7 above)
Comment (hudaalamassi) [reply solutions/num_converter.py]: Consider adding an assertion message to clarify the purpose of the check. This will make it easier to debug if the assertion fails by providing context about what went wrong."

 Example:
 
- Instead of: assert isinstance(decimal, int)

- You could write: assert isinstance(decimal, int), "Input must be an integer"
- This way, if the assertion fails, the error message "Input must be an integer" will appear, making it clear why the check failed.

File: solutions/num_converter.py (hunk 10 above)
Comment (hudaalamassi) [reply solutions/num_converter.py:42]: this my mistake, nothing wrong here, I checked my understanding of the term and found that I did not fully understand it. I thought that all variables should be named with multiple words to follow snake_case

File: solutions/num_converter.py (hunk 8 above)
Comment (hudaalamassi) [reply solutions/num_converter.py:61]: I checked my understanding of the term and found that I did not fully understand it. I thought that all variables should be named with multiple words to follow snake_case

File: solutions/num_converter.py (hunk 11 above)
Comment (hudaalamassi) [reply solutions/num_converter.py:93]: I checked my understanding of the term and found that I did not fully understand it. I thought that all variables should be named with multiple words to follow snake_case

File: solutions/num_converter.py (hunk 8 above)
Comment (hudaalamassi) [reply solutions/num_converter.py:61]: You're correct that implementation comments describe the specific code. My suggestion is aimed at shifting the focus of the comment from the how to the why. While your comment describes the action, adding the reasoning behind it (reversing the digits to produce the final binary string) makes it more strategy-oriented. This gives future readers insight into the purpose of the step.

File: solutions/num_converter.py (hunk 7 above)
Comment (byabdallah) [reply solutions/num_converter.py]: Done! Thanks for the hint.

File: solutions/num_converter.py (hunk 4 above)
Comment (hudaalamassi) [reply solutions/num_converter.py:72]: In the docstring: (as a string or integer), I think string should be removed

File: solutions/num_converter.py (hunk 3 above)
Comment (hudaalamassi) [reply solutions/num_converter.py:21]: I think to follow the rule completely for other assumptions, the docstring for parameter should show expected range of the integer (e.g., if it should be non-negative).

File: solutions/num_converter.py (hunk 4 above)
Comment (byabdallah) [reply solutions/num_converter.py:72]: I did change that but not quite sure why it's still the same here

File: solutions/num_converter.py (hunk 3 above)
Comment (byabdallah) [reply solutions/num_converter.py:21]: This was already illustrated in the module docstring, no need to repeat it every couple of lines I believe

Comment (hudaalamassi) [reply solutions/num_converter.py:21]: Screenshot_20250104_183627_com_android_chrome_ChromeTabbedActivity_edit_17998959023133

I see your point, but there is a check for this in the checklist

Comment (byabdallah) [reply solutions/num_converter.py:21]: Yes the function behavior is described
image

Comment (hudaalamassi) [reply solutions/num_converter.py:21]: yes you are right but the behavior point is the main, and I couldn't check it, because the other assumptions not clear in the code from my point of view, also the type I do not know why the other function argument is still containing str, so I couldn't check the type

Comment (byabdallah) [reply solutions/num_converter.py:21]: What are the assumptions that are not clear? Which argument contains str?

Comment (hudaalamassi) [reply solutions/num_converter.py:21]: Screenshot_20250104_190747_com_android_chrome_ChromeTabbedActivity_edit_20062679161617
Screenshot_20250104_190444_com_android_chrome_ChromeTabbedActivity_edit_20047516683758
 

This is about the type, I do not know why in a place the type is corrected and in other it is not appeared to be corrected. But I think the type will consider resolved now.

Comment (hudaalamassi) [reply solutions/num_converter.py:21]: - For other assumptions, the rule is in the function docstring.

- The description for non-negative is in the module docstring now, so i think this addition in the function docstring is good practice to add the expected range for the parameters.


- like this for example: 

        decimal_number (int): The decimal number to be converted. 
                               It is assumed that the decimal number is a non-negative integer

**Review by hudaalamassi — APPROVED** [review]
The code is perfect! Well done!
